### DISCUSSION TITLE
emqx:4.3.0 exhook grpc too slow？

### DISCUSSION THREAD

**Discussion by Bevr2020** [emqx/emqx#5218] [Q&A]
**Please describe your problem in detail, if necessary, you can upload the log file through the attachment**:
when i start emqx exhook plugins using java grpc, how can i improve the grpc call method(like onClientConnected, onClientAuthenticate) speed? how can i configure the grpc client ? grpc server has configured threadpool, however now the method call speed is 20 times per second at most, Too Slow.

help me plz. thx

**Comment by k32** [comment]
Hello,

Please share the following information:
1) Do you run any business logic in the request handler?
2) Which libraries do you use? Which async primitives do you use? 
Writing a gRPC server in Java can be done in a number of different ways, we need to know more about your specific case. Can you share some code of your service, etc.
3) Can you share some characteristics of the hardware that you're using.

**Reply by Bevr2020** [reply]
![微信图片_20210714211538](https://user-images.githubusercontent.com/62202698/125627605-e220279d-e704-4c54-a867-9dcd10610a94.png)
![微信图片_20210714211737](https://user-images.githubusercontent.com/62202698/125627892-1ee62243-3a70-48c9-889b-b5c08475bdf9.jpg)
![微信图片_20210714212152](https://user-images.githubusercontent.com/62202698/125628456-381b5af0-75a8-4c37-b449-7f5273ff345e.jpg)

1: i do not run any business logic
2: as exhook logs show, the grpc callback is slow, nearly 50-100ms between method callback, is there any configure for exhook grpc client threadpool ? 
this is what i configure in exhook grpc thread pool, but

**Reply by hjianbo** [reply]
In our internal tests, exhook can achieve a request speed of roughly 20k/s (on a 16-core virtual machine)

I think it is difficult to evaluate exhook's performance metrics with the logging interval approach. Because the requests are sequential and network delayed for a single client, a few tens of milliseconds between requests is quite reasonable.

**Reply by hjianbo** [reply]
By default, the size of exhook's thread pool is equal to the number of CPUs. There is no available configuration out there yet.

You can try [emqtt_bench conn](https://github.com/emqx/emqtt-bench) to create a large number of connections at the same time to get the maximum number of concurrent connections it can support per second.

**Reply by jia-zhengwei** [reply]
> ![微信图片_20210714211538](https://user-images.githubusercontent.com/62202698/125627605-e220279d-e704-4c54-a867-9dcd10610a94.png) ![微信图片_20210714211737](https://user-images.githubusercontent.com/62202698/125627892-1ee62243-3a70-48c9-889b-b5c08475bdf9.jpg) ![微信图片_20210714212152](https://user-images.githubusercontent.com/62202698/125628456-381b5af0-75a8-4c37-b449-7f5273ff345e.jpg)
> 
> 1: i do not run any business logic 2: as exhook logs show, the grpc callback is slow, nearly 50-100ms between method callback, is there any configure for exhook grpc client threadpool ? this is what i configure in exhook grpc thread pool, but

Hi, Bro 
Do you have any suggestion on this question?   **the grpc callback is  really slow!**
